flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,442 @@
+
+    // Initialize the timer states.
+    val leftTimerStateDesc: ValueStateDescriptor[Long] =
+      new ValueStateDescriptor[Long](timeIndicator + "InnerJoinLeftTimerState", classOf[Long])
+    leftTimerState = getRuntimeContext.getState(leftTimerStateDesc)
+
+    val rightTimerStateDesc: ValueStateDescriptor[Long] =
+      new ValueStateDescriptor[Long](timeIndicator + "InnerJoinRightTimerState", classOf[Long])
+    rightTimerState = getRuntimeContext.getState(rightTimerStateDesc)
+  }
+
+  /**
+    * Process rows from the left stream.
+    */
+  override def processElement1(
+      cRowValue: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+    updateOperatorTime(ctx)
+    val rowTime: Long = getTimeForLeftStream(ctx, cRowValue)
+    val oppositeLowerBound: Long = rowTime - rightRelativeSize
+    val oppositeUpperBound: Long = rowTime + leftRelativeSize
+    processElement(
+      cRowValue,
+      rowTime,
+      ctx,
+      out,
+      leftOperatorTime,
+      oppositeLowerBound,
+      oppositeUpperBound,
+      rightOperatorTime,
+      rightTimerState,
+      leftCache,
+      rightCache,
+      leftRow = true
+    )
+  }
+
+  /**
+    * Process rows from the right stream.
+    */
+  override def processElement2(
+      cRowValue: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+    updateOperatorTime(ctx)
+    val rowTime: Long = getTimeForRightStream(ctx, cRowValue)
+    val oppositeLowerBound: Long = rowTime - leftRelativeSize
+    val oppositeUpperBound: Long =  rowTime + rightRelativeSize
+    processElement(
+      cRowValue,
+      rowTime,
+      ctx,
+      out,
+      rightOperatorTime,
+      oppositeLowerBound,
+      oppositeUpperBound,
+      leftOperatorTime,
+      leftTimerState,
+      rightCache,
+      leftCache,
+      leftRow = false
+    )
+  }
+
+  /**
+    * Put a row from the input stream into the cache and iterate the opposite cache to
+    * output join results meeting the conditions. If there is no timer set for the OPPOSITE
+    * STREAM, register one.
+    */
+  private def processElement(
+      cRowValue: CRow,
+      timeForRow: Long,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow],
+      myWatermark: Long,
+      oppositeLowerBound: Long,
+      oppositeUpperBound: Long,
+      oppositeWatermark: Long,
+      oppositeTimeState: ValueState[Long],
+      rowListCache: MapState[Long, JList[Row]],
+      oppositeCache: MapState[Long, JList[Row]],
+      leftRow: Boolean): Unit = {
+    cRowWrapper.out = out
+    val row = cRowValue.row
+    if (!checkRowOutOfDate(timeForRow, myWatermark)) {
+      // Put the row into the cache for later use.
+      var rowList = rowListCache.get(timeForRow)
+      if (null == rowList) {
+        rowList = new ArrayList[Row](1)
+      }
+      rowList.add(row)
+      rowListCache.put(timeForRow, rowList)
```

Comment:
Just had a discussion about this with a colleague. 

He suggested to round the timestamp to reduce the number of keys in the state backend. This would also mean that we can directly address all keys (because we can compute them) that we need to join with and don't need to iterate over all keys. Clean-up would also work without full traversal.

However, when joining we would need to check again the window condition because a list might contain records that have the same rounded timestamp but are outside of the window. For that we need to store the rows with timestamps (at least for proctime, rowtime has the key already in the row).

This design should give much better performance because we only access relevant keys. However, we would need to decide for a key granularity. I think a few seconds (10 seconds) could be a good starting point.

What do you think about this @xccui?

Reply:
Hi @fhueske, that's a good idea for me! We could start from a constant granularity and make it dynamic in the future.
BTW, I'm not sure if you still remember the following comment in FLINK-6233 😄 
> To cope with that, I plan to split the "cache window" into continuous static-panes, and casting one to expired as a whole. By doing like that, we may store some extra records, whose time interval is equal to the static span of the panes, but can remove the expired data efficiently.

Reply:
Ah, I misread your proposal then :-) I think starting from a constant value is a good idea. When making it dynamic we have to be careful. Too high granularity will lead to larger lists and higher ser/de overhead.

Btw. while iterating over the list of a rounded key, we can still clean-up the state using the same condition as before and eagerly reduce the state size.